### PR TITLE
build: api point to display knowledge panels

### PR DESCRIPTION
**Pull request by sumit-158**
### What
- Small javascript function to display knowledge panels built with [jinja2](https://fastapi.tiangolo.com/advanced/templates/)
- Able to display single and multiple panels
 
### Fixes bug(s)
- #48

### CODE REVIEW

**Comment by teolemon** [comment]
cc @CharlesNepote it would be great to include that in https://github.com/openfoodfacts/power-user-script

File: template/knowledge_panels.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    var stack1, lookupProperty = container.lookupProperty || function(parent, propertyName) {
+        if (Object.prototype.hasOwnProperty.call(parent, propertyName)) {
+          return parent[propertyName];
+        }
+        return undefined
+    };
+
+  return ((stack1 = lookupProperty(helpers,"if").call(depth0 != null ? depth0 : (container.nullContext || {}),(depth0 != null ? lookupProperty(depth0,"knowledge_panels") : depth0),{"name":"if","hash":{},"fn":container.program(1, data, 0),"inverse":container.noop,"data":data,"loc":{"start":{"line":1,"column":0},"end":{"line":17,"column":7}}})) != null ? stack1 : "");
+},"useData":true});
+})();
```
Comment (alexgarel) [inline template/knowledge_panels.js:79]: What is this file ? Is that the compilation of the handlebars template ? Why would we need to have this ?

**Review by alexgarel — CHANGES_REQUESTED** [review]
Could you add a FastAPI route that renders this html ?
```
/render-to-html/?{same parameters as knowledge panels api}
```

File: template/knowledge_panels.js (hunk 1 above)
Comment (sumit-158) [reply template/knowledge_panels.js:79]: Yes, Its Precompiling template made by the handlebar itself for "knowledge_panels.handlebars"

Comment (sumit-158) [reply template/knowledge_panels.js:79]: https://handlebarsjs.com/installation/precompilation.html#getting-started

Comment (sumit-158) [reply template/knowledge_panels.js:79]: we can do something like this too, then we don't need Precompiling template
image

Comment (alexgarel) [reply template/knowledge_panels.js:79]: Yes, I would avoid this pre-compiling. It's just a small tool for now.

File: template/index.html [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    <script type="text/javascript">
+      // Assign the template name
+      var source = document.getElementById("entry-template").innerHTML;
+      var template = Handlebars.compile(source);
+
+      // Compile the template
+      var context = {
+        // Response from the API
+      };
```
Comment (sumit-158) [inline template/index.html:42]: @alexgarel question: How I can inject the API response I.e, "/knowledge_panel?{query}" into the "context"

**Comment by sumit-158** [comment]
Ping @alexgarel

File: app/main.py [hunk 3]
```diff
@@ -112,3 +118,53 @@ async def knowledge_panel(
...
+    lang_code: Union[str, None] = Query(
+        default=None,
+        title="language code string",
+        description="To return knowledge panels in native language, defualt lang: `en`.",
+    ),
+    country: Union[str, None] = Query(
+        default=None,
+        title="Country tag string",
+        description="To return knowledge panels for specific country, ex: `france`.",
+    ),
```
[the remark this answers is not in the document]
Comment (alexgarel) [reply app/main.py:157]: We could at least put the "Query" in a dataclass to reuse them.

File: app/main.py [hunk 4]
```diff
@@ -112,3 +118,53 @@ async def knowledge_panel(
                 logging.exception()
 
         return {"knowledge_panels": panels}
+
+
+templates = Jinja2Templates(directory="template")
+
+
+@app.get("/render-to-html", tags=["Render to HTML"], response_class=HTMLResponse)
+async def render_html(
```
Comment (alexgarel) [inline app/main.py:127]: I think we should have a parameter to say if we want a full html or just the "entry" div (for inclusion on an existing page).

File: template/item.html [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<html>
+  <head>
+    <title>Panel Details</title>
+  </head>
+  <body>
+    <div class="entry">
```
Comment (alexgarel) [inline template/item.html:6]: I would use a very specific class to ease integration on another site.

```
div class="knowledge-facts-panel entry"
```

File: app/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sumit-158) [reply app/main.py]: Is the Current approach is good?

File: app/main.py (hunk 4 above)
Comment (sumit-158) [reply app/main.py:104]: I just removed Html and body tags because even for full HTML rendering HTML tags aren't essential.

**Review by alexgarel — APPROVED** [review]
Sumit, I tested it and while at it made some improvements:

* wrap each panel in a div with panel name as class
* link to product in last edits
* use a thumbnail for wikidata image

**Comment by sumit-158** [comment]
Thank you! @alexgarel for the review and improvement :-)

**Pull request merged**